compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -2516,15 +2521,18 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
     case _ =>
       tp
 
-  private def andTypeGen(tp1: Type, tp2: Type, op: (Type, Type) => Type,
-      original: (Type, Type) => Type = _ & _, isErased: Boolean = ctx.erasedTypes): Type = trace(s"andTypeGen(${tp1.show}, ${tp2.show})", subtyping, show = true) {
+  private def andTypeGen(tp1orig: Type, tp2orig: Type, op: (Type, Type) => Type,
+      original: (Type, Type) => Type = _ & _, isErased: Boolean = ctx.erasedTypes): Type = trace(s"andTypeGen(${tp1orig.show}, ${tp2orig.show})", subtyping, show = true) {
```

Comment:
I think it's clearer and more efficient like this:
```scala
tp1 match
  case FlexibleType(_, hi1) => FlexibleType(andTypeGen(hi1, tp2))
  tp2 match 
    case FlexibleType(_, hi2) => FlexibleType(andTypeGentp1, hi2))
    case _ => ...
```

Reply:
Generally, avoid using the unapply of FlexibleType since that can be a performance issue.

Reply:
Sorry I forget to reply to this comment: the original code returns a flexible type if **both** types are flexible types; the suggested code returns a flexible type if **one of** the types is a flexible type. The original code can avoid a special case where flexible types are keep adding to the result type when the original type is returned.

Reply:
Now, I think both versions are not correct. I will think about how to modify these functions.

Reply:
Unnecessary code removed